### ISSUE TITLE
Breakpoints broken via Attach to running NodeJS process (broken in 0.10.5 and 0.10.6)

### ISSUE THREAD

**Issue by igorvaschuk** [Microsoft/vscode#1552]
Updated on two machines and on both machines breakpoints are no longer getting hit, despite Code appearing to be attached to the process.
Runing process via "nodemon --debug ./bin/www", then attaching via "Attach" in VS Code.
Debugging via "Launch" works fine and breakpoints are hit.
OS: Windows 10
Node version: v4.2.2
Project created via Visual Studio 2015 with NTVS -> New Project->Templates->Other Languages->JavaScript->Basic Node.js Express 4 Application

**Comment by bpasero** [comment]
@igorvaschuk which nodejs version?

**Comment by weinand** [comment]
@igorvaschuk I've installed MEAN and created a first app with "mean init yourNewApp". Since I'm unfamiliar with MEAN could you please guide me how to create an app that lives in ./bin/www so that I can reproduce the problem.

BTW, I could not reproduce the attach problem with an express app.

**Comment by gclifford** [comment]
I was also getting this when running on Ubuntu. I was trying to attach to a standard expressjs app and it would not hit any of the break point. If I used Launch it worked fine.

**Comment by weinand** [comment]
@igorvaschuk @gclifford could you please try to use a non-default debug port for the express app, e.g. something like "node --debug=5555 bin/www" and then modify the launch config in VSCode to the same port. In my experiments I've seen a problem with the standard 5858 debug port.

**Comment by jpierson** [comment]
I'm noticing that I'm having problems lately hitting breakpoints in VSCode for my node project as well. Somewhat recently [nodemon](https://github.com/remy/nodemon/) was added to the mix via [gulp-nodemon](https://github.com/JacksonGariety/gulp-nodemon) where gulp was already used for launching the debug mode of the app. I'm guessing that there may be something with how _nodemon_ works which is causing problems for VSCode. Not sure where or if this can be addressed.

Possible realated issues:

* https://github.com/remy/nodemon/issues/546
* https://github.com/JacksonGariety/gulp-nodemon/issues/92
* https://github.com/JacksonGariety/gulp-nodemon/issues/85

**Comment by jpierson** [comment]
@weinand I've tried your advice with the debug port but it didn't seem to help in my case.

**Comment by gclifford** [comment]
@weinand I am not experiencing this anymore. I am wondering if there was a difference between the TypeScript code and sourceMaps and the debugger got confused.

**Comment by bartvanderwal** [comment]
:+1:! This is really vital
As Ward Bell stated in the comment on [John Papa's blog on debugging in VS Code](http://johnpapa.net/debugging-with-visual-studio-code/): "Whether client or server side. Auto-refresh as we do with nodemon should be part of the solution."

I switched over from Visual Studio NTVS to Visual Studio Code because of the terrible stalling of the UI in VS 2015 while editing typescript. VS Code is a lot more workable, but live debugging is now still not possible.

**Comment by weinand** [comment]
@bartvanderwal I suggest that you file a separate issue for your feature request because this issue is a bug.

**Comment by bartvanderwal** [comment]
Thanks for the quick reply @weinand, but actually it IS working now. Somewhat. At least I won't post an issue.

In case this might help anyone It's working with below settings in Current VS Code version (e.g. v0.10.6) on Windows 10. But it'll probably work on Mac too (I might check later).

`/.vscode/launch.json`
```json
{
	"configurations": [{
		"name": "Launch",
		"outDir": null

	},{
		"name": "Attach",
		"type": "node",
		"request": "attach",
		"port": 5858
	}]
}
```

For completeness I'm running Nodemon using grunt with below settings, but the equivalent command line should do the same.

`/Gruntfile.js`
```json
nodemon : {
	dev : {
	script : 'launcher.js'
	},
	options : {
		ignore : ['node_modules/**', 'Gruntfile.js'],
                nodeArgs: ['--debug'],
		env : { PORT : '4123'
		}
	}
}
```

I guess the debug port `5858` is the default since it's not specified here (but it is in `launch.json` above.)

**Comment by weinand** [comment]
If breakpoints in your startup code are not hit in "attach" mode, but they work in "launch" mode please try to start your target with the "--debug-brk" option instead of the "--debug" option.

The "--debug-brk" option makes sure that the target stops on entry and that VS Code gets a chance to  properly register the breakpoints. When starting with "--debug" only, the debugger runs over the startup code before VS Code has a chance to register the breakpoints.

In "launch" mode it is not necessary to set `stopOnEntry` to true in order to get the breakpoint in startup code hit because VS Code behind the scenes always launches the debugger with "--debug-brk" (but hides the break on entry to the user).

**Comment by weinand** [comment]
@bartvanderwal I've created this feature-request for the automatic reattach:  https://github.com/Microsoft/vscode/issues/2103

**Comment by bartvanderwal** [comment]
That looks great @weinand! I'll try it out shortly.

**Comment by ianwesterfield** [comment]
I have found one weird workaround with node --debug (default port) and node-inspector.

Same-module breakpoints in vscode begin working as expected after setting and hitting just one breakpoint using node-inspector in that module. Breakpoints set by vscode in other modules fail to hit until node-inspector first hits a breakpoint in the same module.

**Comment by alexisbg** [comment]
My Express 4 application is experiencing a similar issue.

It runs on a Debian 8 box, Node.js version is 4.3.0 and VS Code 0.10.8 runs on Windows 7. I start it with:
`node --debug-brk server/server.js`

VS Code breaks at the first line. Next every breakpoints in server/server.js are correctly handled by VS Code. But all the breakpoints ouside this file will never stop.

On VS Code Debug view, server/server.js is marked as "content streamed from server" but not on Exporer view. By the way its breakpoints are only visible in the Debug view. It seems that only this file is processed by VS Code.

On the other hand, Node Inspector works like a charm:
`node-debug --web-host=192.168.101.102 server/server.js`

**Comment by alexisbg** [comment]
VS Code has been updates to v0.10.10 but there is still the same behaviour.

After reflection, I figured out that VS Code was not able to map remote files to those of the Explorer view. I investigated the vscode-debug-node project and I have found the localRoot and remoteRoot parameters of attach configuration:
[https://github.com/Microsoft/vscode-node-debug/blob/master/src/node/nodeDebug.ts#L146](url)

And there is a short description about them on January release page, below "Remote Debugging":
[https://code.visualstudio.com/updates/vJanuary#_nodejs-debugging](url)

These 2 parameters fix the issue. Here is a launch.json configuration sample:

```json
{
    "name": "Attach to my.remote.linux.box",
    "type": "node",
    "request": "attach",
    "address": "my.remote.linux.box",
    "port": 5858,
    "localRoot": "${workspaceRoot}",
    "remoteRoot": "/path/to/project/root/on/remote/host"
}
```

I hope this can help.
